services/src/main/java/org/keycloak/services/clientpolicy/DefaultClientPolicyManager.java

```diff
@@ -17,86 +17,90 @@
-        if (!Profile.isFeatureEnabled(Profile.Feature.CLIENT_POLICIES)) return;
-        ClientPolicyLogger.logv(logger, "Client Policy Operation : event = {0}", context.getEvent());
+        if (!Profile.isFeatureEnabled(Profile.Feature.CLIENT_POLICIES)) {
+            return;
+        }
+
+        RealmModel realm = session.getContext().getRealm();
+        ClientPolicyLogger.logv(logger, "{0} :: POLICY OPERATION :: context realm = {1}, event = {2}", logMsgPrefix(), realm.getName(), context.getEvent());
+
         doPolicyOperation(
                 (ClientPolicyConditionProvider condition) -> condition.applyPolicy(context),
-                (ClientPolicyExecutorProvider executor) -> executor.executeOnEvent(context)
-            );
+                (ClientPolicyExecutorProvider executor) -> executor.executeOnEvent(context),
+                realm
+        );
     }
 
-    private void doPolicyOperation(ClientConditionOperation condition, ClientExecutorOperation executor) throws ClientPolicyException {
-        RealmModel realm = session.getContext().getRealm();
-        for (ClientPolicyProvider policy : getProviders(realm)) {
-            ClientPolicyLogger.logv(logger, "Policy Operation : name = {0}, provider id = {1}", policy.getName(), policy.getProviderId());
-            if (!isSatisfied(policy, condition)) continue;
-            execute(policy, executor);
-        }
-    }
-
-    private List<ClientPolicyProvider> getProviders(RealmModel realm) {
-        List<ClientPolicyProvider> providers = providersMap.get(realm.getId());
-        if (providers == null) {
-            providers = realm.getComponentsStream(realm.getId(), ClientPolicyProvider.class.getName())
-                    .map(policyModel -> {
-                        try {
-                            ClientPolicyProvider policy = session.getProvider(ClientPolicyProvider.class, policyModel);
-                            ClientPolicyLogger.logv(logger, "Loaded Policy Name = {0}", policyModel.getName());
-                            session.enlistForClose(policy);
-                            return policy;
-                        } catch (Throwable t) {
-                            logger.errorv(t, "Failed to load provider {0}", policyModel.getId());
-                            return null;
-                        }
-                    })
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toList());
-            providersMap.put(realm.getId(), providers);
-        } else {
-            ClientPolicyLogger.log(logger, "Use cached policies.");
+    private void doPolicyOperation(ClientConditionOperation condition, ClientExecutorOperation executor, RealmModel realm) throws ClientPolicyException {
+        Map<String, ClientProfileModel> map = ClientPoliciesUtil.getClientProfilesModel(session, realm);
+        List<ClientPolicyModel> list = ClientPoliciesUtil.getClientPoliciesModel(session, realm).stream()
```

Comment:
Method `ClientPoliciesUtil.getClientProfilesModel` can be renamed to `getEnabledClientProfilesModel` and the filtering for enabled can be done directly in this method. No need for additional filtering here then. WDYT?

Reply:
@mposolda I agree with you. I will filter enabled Profiles by renamed `ClientPoliciesUtil.getEnabledClientProfilesModel` and remove this filtering operation in `DefaultClientPolicyManager.doPolicyOperation` that calls this method.

Reply:
Thanks